### PR TITLE
Java: Add query for improper webview certificate validation

### PR DESCRIPTION
**Pull request by joefarebrother**
This query finds implentations of `WebViewClient.onRecievedSslError` that unconditionaly call `SslErrorHandler.proceed` and thus accept all certificates.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/security/AndroidWebViewCertificateValidationQuery.qll [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/** Defintions for the web view certificate validation query */
```
Comment (atorralba) [inline java/ql/lib/semmle/code/java/security/AndroidWebViewCertificateValidationQuery.qll:1]:
```suggestion
/** Definitions for the web view certificate validation query */
```

File: java/ql/lib/semmle/code/java/security/AndroidWebViewCertificateValidationQuery.qll [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+/** A call to `SslErrorHandler.proceed` */
+private class SslProceedCall extends MethodAccess {
+  SslProceedCall() {
+    this.getMethod().hasQualifiedName("android.webkit", "SslErrorHandler", "proceed")
+  }
+}
+
+/** Holds if `m` trusts all certifiates by calling `SslErrorHandler.proceed` unconditionally. */
```
Comment (atorralba) [inline java/ql/lib/semmle/code/java/security/AndroidWebViewCertificateValidationQuery.qll:31]:
```suggestion
/** Holds if `m` trusts all certificates by calling `SslErrorHandler.proceed` unconditionally. */
```

File: java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertifiacteValidation.qhelp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  <li>The vulnerable application calls the <code>onReceivedSslError</code> method to check whether it should trust the certificate.</li>
+  <li>The <code>onReceivedSslError</code> method of your <code>WebViewClient</code> calls <code>SslErrorHandler.proceed</code>.</li>
+  <li>The vulnerable application accepts the certificate and proceeds with the connection since your <code>WevViewClient</code> trusted it by proceeding.</li>
+  <li>The attacker can now read the data your application sends to <code>https://example.com</code> and/or alter its replies while the application thinks the connection is secure.</li>
+</ol>
+</overview>
+
+<recommendation>
+<p>
+Do not use a call <code>SslerrorHandler.proceed</code> unconditonally.
```
Comment (atorralba) [inline java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertifiacteValidation.qhelp:27]:
```suggestion
Do not use a call <code>SslerrorHandler.proceed</code> unconditionally. 
```

File: java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.ql [hunk 4]
```diff
@@ -0,0 +1,18 @@
+/**
+ * @name Android `WebVeiw` that accepts all certificates
```
Comment (atorralba) [inline java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.ql:2]:
```suggestion
 * @name Android `WebView` that accepts all certificates
```

File: java/ql/src/change-notes/2022-06-22-improper-webview-certificate-validation.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+---
+category: newQuery
+---
+* A new query "Android `WebVeiw` that accepts all certificates" (`java/improper-webview-certificate-validation`) has been added. This query finds implementations of `WebViewClient`s that accept all certificates in the case of an SSL error.
```
Comment (atorralba) [inline java/ql/src/change-notes/2022-06-22-improper-webview-certificate-validation.md:4]:
```suggestion
* A new query "Android `WebView` that accepts all certificates" (`java/improper-webview-certificate-validation`) has been added. This query finds implementations of `WebViewClient`s that accept all certificates in the case of an SSL error.
```

**Review by atorralba — COMMENTED** [review]
Some quick comments, I'll do a more detailed review later.

File: java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.qhelp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+<example>
+<p>
+In the first (bad) example, the <code>WebViewClient</code> trusts all certificates by always calling <code>SslErrorHandler.proceed</code>. 
+In the second (good) example, only certificates signed by a certain public key are accepted.
+</p>
+<sample src="ImproperWebViewCertificateValidation.java" />
+</example>
+
+<references>
```
Comment (atorralba) [inline java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.qhelp:41]: Suggested reference: https://developer.android.com/reference/android/webkit/WebViewClient?hl=en#onReceivedSslError(android.webkit.WebView,%20android.webkit.SslErrorHandler,%20android.net.http.SslError)

**Review by atorralba — COMMENTED** [review]
Let's go for a performance evaluation and LGTM run for FP ratio. But otherwise looks good to me, just added a minor suggestion.

File: java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.ql [hunk 7]
```diff
@@ -0,0 +1,18 @@
+/**
+ * @name Android `WebView` that accepts all certificates
+ * @description Trusting all certificates allows an attacker to perform a machine-in-the-middle attack.
```
Comment (mchammer01) [inline java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.ql:3]: Sorry, commenting here to check that `machine-in-the-middle attack` is a thing (I've only ever heard of `man-in-the-middle` attacks 🙊 ).

File: java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.qhelp [hunk 8]
```diff
@@ -0,0 +1,46 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>
+If the <code>onReceivedSslError</code> method of an Android <code>WebViewClient</code> always calls <code>proceed</code> on the given <code>SslErrorHandler</code>, it trusts any certificate. 
+This allows an attacker to perform a machine-in-the-middle attack against the application, therefore breaking any security Transport Layer Security (TLS) gives.
```
Comment (mchammer01) [inline java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.qhelp:8]: Same comment as on the ql file (double-checking)

**Review by mchammer01 — DISMISSED** [review]
@joefarebrother 👋🏻 - I carried out an editorial review and this LGTM ✨ 
Just a couple of clarifying comments.

File: java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.ql (hunk 7 above)
Comment (joefarebrother) [reply java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.ql:3]: It should be man-in-the-middle, thanks

Comment (atorralba) [reply java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.ql]: I thought you used `machine-in-the-middle` as a gender-neutral term :-)

Comment (joefarebrother) [reply java/ql/src/Security/CWE/CWE-295/ImproperWebViewCertificateValidation.ql]: In fact I copied it from another file. But that reasoning makes sense; I'll keep it as machine then

**Comment by joefarebrother** [comment]
Is this ready to merge? @atorralba

**Review by atorralba — APPROVED** [review]
Code and LGTM results look good to me. If DCA is happy, I think we can merge! :)

**Pull request merged**